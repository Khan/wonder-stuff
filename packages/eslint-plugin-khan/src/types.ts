// We need to define our own PluginDocs type in order for us to be
// able to set the `recommended` field in meta.docs when calling
// `createRule`.  For more info, see:
// https://typescript-eslint.io/developers/custom-rules/#extra-rule-docs-types
export interface MyPluginDocs {
    recommended: boolean;
}
