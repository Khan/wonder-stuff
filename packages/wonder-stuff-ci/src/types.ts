/**
 * The information about a release branch.
 */
export type MobileReleaseBranchInfo = {
    prefix: string;
    version: string;
};

/**
 * The options for the Winston GCP transport.
 */
export type GCPTransportOptions = {
    /**
     * The GCP project ID to send logs to.
     */
    projectId: string;
    /**
     * The name of the log to send logs to in GCP.
     */
    logName: string;
    /**
     * If true, logs will be written to stdout instead of GCP
     */
    redirectToStdout: boolean;
    /**
     * The log level to send to GCP.
     */
    level: string;
    /**
     * K/V labels that will be add to the logs
     */
    labels: {[key: string]: string};
    /**
     * K/V for setting the resource values for GCP
     */
    resource: {[key: string]: string};
};

/**
 * The gcp log levels that are supported.
 */
export type GCPLogLevels = "debug" | "info" | "warn" | "error";

/**
 * The options for the mobile release logger.
 */
export type MobileReleaseLoggerOptions = {
    /**
     * If true, logs will be written to stdout.
     */
    redirectToStdout?: boolean;
    /**
     * K/V labels that will be add to the logs
     */
    labels?: {[key: string]: string};
    /**
     * The minimum log level to send to GCP.
     */
    logLevel?: GCPLogLevels;
    /**
     * The name of the log to send logs to in GCP.
     */
    logName?: string;
    /**
     * The GCP project ID to send logs to.
     */
    projectId?: string;
    /**
     * K/V metadata that will be add to each the log
     */
    defaultMetadata?: {[key: string]: string};
    /**
     * K/V for setting the resource values for GCP, where resource.type is required by GCP API.
     */
    resource?: {[key: string]: string};
};
