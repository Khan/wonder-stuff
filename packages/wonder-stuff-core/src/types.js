// @flow
export type {ErrorKind} from "./errors.js";

/**
 * A collection of data.
 */
export type Metadata = {
    +[name: string]: Metadata | string | number | boolean | null,
};
