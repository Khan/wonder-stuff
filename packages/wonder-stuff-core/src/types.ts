type MetadataPrimitive = string | number | boolean | null | undefined;
type MetadataArray<T> = Array<T | MetadataArray<T>>;

/**
 * A collection of data.
 */
export type Metadata = {
    [name: string]:
        | Metadata
        | MetadataPrimitive
        | MetadataArray<MetadataPrimitive | Metadata>;
};
