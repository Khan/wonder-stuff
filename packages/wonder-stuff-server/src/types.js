// @flow
import type {
    NpmLogLevels,
    Logger as WinstonLogger,
    Info as WinstonInfo,
    Transport,
} from "winston";

import type {$Request} from "express";

/**
 * Describes logging metdata.
 */
export type Info = WinstonInfo<NpmLogLevels>;

/**
 * Defines the different log levels.
 */
export type LogLevel = $Keys<NpmLogLevels>;

/**
 * Describes the interface for logging gateway activity.
 */
export type Logger = WinstonLogger<NpmLogLevels>;

/**
 * The runtime modes that a gateway can run under.
 */
export enum Runtime {
    // TODO(somewhatabstract, FEI-4172): Update eslint-plugin-flowtype when
    // they've fixed https://github.com/gajus/eslint-plugin-flowtype/issues/502
    /* eslint-disable no-undef */
    Production = "production",
    Development = "development",
    Test = "test",
    /* eslint-enable no-undef */
}

/**
 * Options to configure logging.
 */
export type LoggingOptions = {
    /**
     * The runtime mode to be targeted. When omitted, this will default to
     * the current runtime mode as determined from `getRuntimeMode`.
     */
    mode?: Runtime,

    /**
     * Log only if the level of a logged entry is less than or equal to this
     * level. Enables filtering out of debug message in production, for example.
     */
    level: LogLevel,

    /**
     * Metadata to attach to every logged entry, by default.
     * Defaults to null.
     */
    defaultMetadata?: {...} | (() => {...}),

    /**
     * The transport to use for handling log entries. When not specified,
     * this will be determined using `getLoggingTransport`.
     */
    transport?: Transport,
};

/**
 * Options to configure a server.
 */
export type ServerOptions = {
    /**
     * The name of the server.
     */
    +name: string,

    /**
     * The port on which the server should listen.
     */
    +port: number,

    /**
     * The hostname to which the server should bind.
     */
    +host: string,

    /**
     * The logger to use for logging.
     */
    +logger: Logger,

    /**
     * What runtime mode the server is running under.
     */
    +mode: Runtime,

    /**
     * Optional value in milliseconds for keepalive timeout of the server.
     * For running in Google Cloud, this should be higher than the load
     * balancer's own keepalive timeout value, which at time of writing was
     * indicated to be 80000ms [1].
     *
     * [1] https://khanacademy.slack.com/archives/CJSE4TMQX/p1573252787333500
     *
     * Defaults to 90000.
     */
    +keepAliveTimeout?: number,

    /**
     * When `true`, the "heapdump" package will be loaded, allowing for
     * heap dumps to be generated on demand using `kill -USR2 <pid>` where
     * `<pid>` is the process ID of the server.
     *
     * Defaults to `false` in production mode, and `true` in all other modes.
     * Set explicitly to `false` to disable heap dumps in all modes.
     */
    +allowHeapDumps?: boolean,
};

export type RequestWithLog<TReq: $Request> = TReq & {
    log?: Logger,
};
