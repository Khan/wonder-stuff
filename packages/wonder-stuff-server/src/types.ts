// @ts-expect-error [FEI-5011] - TS2305 - Module '"winston"' has no exported member 'NpmLogLevels'. | TS2724 - '"winston"' has no exported member named 'Info'. Did you mean 'info'? | TS2724 - '"winston"' has no exported member named 'Transport'. Did you mean 'transport'?
import type {
    NpmLogLevels,
    Logger as WinstonLogger,
    Info as WinstonInfo,
    Transport,
} from "winston";

import type {Request} from "express";

/**
 * Describes logging metdata.
 */
export type Info = WinstonInfo<NpmLogLevels>;

/**
 * Defines the different log levels.
 */
export type LogLevel = keyof NpmLogLevels;

/**
 * Describes the interface for logging gateway activity.
 */
// @ts-expect-error [FEI-5011] - TS2315 - Type 'Logger' is not generic.
export type Logger = WinstonLogger<NpmLogLevels>;

/**
 * The runtime modes that a gateway can run under.
 */
// TODO(FEI-5001): Replace with TS enum
export const Runtime = {
    Production: "production" as const,
    Development: "development" as const,
    Test: "test" as const,
} as const;

/**
 * Options to configure logging.
 */
export type LoggingOptions = {
    /**
     * The runtime mode.
     */
    mode: (typeof Runtime)[keyof typeof Runtime];
    /**
     * Log only if the level of a logged entry is less than or equal to this
     * level. Enables filtering out of debug message in production, for example.
     */
    level: LogLevel;
    /**
     * Metadata to attach to every logged entry, by default.
     * Defaults to null.
     */
    defaultMetadata?: Record<any, any> | (() => Record<any, any>);
    /**
     * The transport to use for handling log entries. When not specified or
     * null, this will be determined using `getLoggingTransport`.
     */
    transport?: Transport | null | undefined;
};

/**
 * Options to configure a server.
 */
export type ServerOptions = {
    /**
     * The name of the server.
     */
    readonly name: string;
    /**
     * The port on which the server should listen.
     */
    readonly port: number;
    /**
     * The hostname to which the server should bind.
     */
    readonly host: string;
    /**
     * The logger to use for logging.
     */
    readonly logger: Logger;
    /**
     * What runtime mode the server is running under.
     */
    readonly mode: (typeof Runtime)[keyof typeof Runtime];
    /**
     * Optional value in milliseconds for keepalive timeout of the server.
     * For running in Google Cloud, this should be higher than the load
     * balancer's own keepalive timeout value, which at time of writing was
     * indicated to be 80000ms [1].
     *
     * [1] https://khanacademy.slack.com/archives/CJSE4TMQX/p1573252787333500
     *
     * Defaults to 90000.
     */
    readonly keepAliveTimeout?: number;
    /**
     * When `true`, the "heapdump" package will be loaded, allowing for
     * heap dumps to be generated on demand using `kill -USR2 <pid>` where
     * `<pid>` is the process ID of the server.
     *
     * Defaults to `false` in production mode, and `true` in all other modes.
     * Set explicitly to `false` to disable heap dumps in all modes.
     */
    readonly allowHeapDumps?: boolean;
    /**
     * When `true` or omitted, the default express-winston-based error logging
     * middleware is included. When `false`, this is not included.
     */
    readonly includeErrorMiddleware?: boolean;
    /**
     * When `true` or omitted, the default express-winston-based request logging
     * middleware is included. When `false`, this is not included.
     */
    readonly includeRequestMiddleware?: boolean;
};

export type RequestWithLog<TReq extends Request> = TReq & {
    log?: Logger;
};
