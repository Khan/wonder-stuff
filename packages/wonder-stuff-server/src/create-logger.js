// @flow
import winston from "winston";
import type {NpmLogLevels} from "winston";

import {Errors} from "./errors.js";
// TODO(somewhatabstract, FEI-4174): Update eslint-plugin-import when they
// have fixed:
// https://github.com/import-js/eslint-plugin-import/issues/2073
// eslint-disable-next-line import/named
import {Runtime} from "./types.js";
import type {LoggingOptions, Logger} from "./types.js";

import {getLoggingTransport} from "./get-logging-transport.js";
import {getRuntimeMode} from "./get-runtime-mode.js";

/**
 * Create a logger with the given options.
 */
export const createLogger = ({
    level,
    defaultMetadata,
    mode,
    transport,
}: LoggingOptions): Logger => {
    mode = mode ?? getRuntimeMode(Runtime.Production);
    const winstonLogger = winston.createLogger<NpmLogLevels>({
        level,
        transports: transport ?? getLoggingTransport(mode, level),
        format: winston.format((info) => {
            // Let's make sure that errors reported without a taxonomic
            // label get labelled.
            if (info.level === "error" && info.kind == null) {
                info.kind = Errors.Internal;
            }
            return info;
        })(),
        defaultMeta: defaultMetadata,
    });

    winstonLogger.debug(
        `Created logger (Level=${level} Mode=${(mode: string)})`,
    );

    return winstonLogger;
};
