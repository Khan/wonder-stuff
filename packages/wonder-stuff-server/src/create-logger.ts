import winston from "winston";

import {Errors} from "./errors";
import type {LoggingOptions, Logger} from "./types";

import {getLoggingTransport} from "./get-logging-transport";

/**
 * Create a logger with the given options.
 */
export const createLogger = ({
    mode,
    level,
    defaultMetadata,
    transport,
}: LoggingOptions): Logger => {
    const winstonLogger = winston.createLogger({
        level,
        transports: transport ?? getLoggingTransport(mode, level),
        format: winston.format((info) => {
            // Let's make sure that errors reported without a taxonomic
            // label get labelled.
            if (info.level === "error" && info.kind == null) {
                info.kind = Errors.Unknown;
            }
            return info;
        })(),
        defaultMeta: defaultMetadata,
    });

    winstonLogger.debug(
        `Created logger (Level=${level} Mode=${mode as string})`,
    );

    return winstonLogger;
};
