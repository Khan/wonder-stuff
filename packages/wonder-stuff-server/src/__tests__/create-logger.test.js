// @flow
import winston from "winston";
import {createLogger} from "../create-logger.js";
// TODO(somewhatabstract, FEI-4174): Update eslint-plugin-import when they
// have fixed:
// https://github.com/import-js/eslint-plugin-import/issues/2073
// eslint-disable-next-line import/named
import {Runtime} from "../types.js";

describe("#createLogger", () => {
    beforeEach(() => {
        // We silence winston's console output.
        jest.spyOn(winston, "createLogger").mockReturnValue({
            debug: jest.fn(),
        });
    });

    describe("during test", () => {
        it("should write to a stream", () => {
            // Arrange
            const mockCreateLogger = jest.spyOn(winston, "createLogger");

            // Act
            createLogger({mode: Runtime.Test, level: "silly"});
            const {transports} = mockCreateLogger.mock.calls[0][0];

            // Assert
            expect(transports).toBeInstanceOf(winston.transports.Stream);
        });

        it("should format log messages to include metadata", () => {
            // Arrange
            jest.spyOn(winston, "createLogger");
            const fakePrintF = jest.fn();
            jest.spyOn(winston.format, "combine", "get").mockReturnValue(
                jest.fn(),
            );
            jest.spyOn(winston.format, "printf", "get").mockReturnValue(
                fakePrintF,
            );

            // Act
            createLogger({mode: Runtime.Test, level: "silly"});
            const result = fakePrintF.mock.calls[0][0]({
                level: "debug",
                message: "MESSAGE",
                other: "metadata",
            });

            // Assert
            expect(result).toMatchInlineSnapshot(`
                "debug: MESSAGE {
                    \\"other\\": \\"metadata\\"
                }"
            `);
        });
    });

    describe("unrecognised runtime mode", () => {
        it("should write to a stream", () => {
            // Arrange
            const mockCreateLogger = jest.spyOn(winston, "createLogger");

            // Act
            createLogger({
                mode: ("MADE UP RUNTIME MODE": $FlowFixMe),
                level: "silly",
            });
            const {transports} = mockCreateLogger.mock.calls[0][0];

            // Assert
            expect(transports).toBeInstanceOf(winston.transports.Stream);
        });
    });

    describe("during development", () => {
        it("should write to the console", () => {
            // Arrange
            const mockCreateLogger = jest.spyOn(winston, "createLogger");

            // Act
            createLogger({mode: Runtime.Development, level: "silly"});
            const {transports} = mockCreateLogger.mock.calls[0][0];

            // Assert
            expect(transports).toBeInstanceOf(winston.transports.Console);
        });

        it.each([undefined, {other: "metadata"}, {}])(
            "should format log messages to include metadata when present (%s)",
            (otherMetadata) => {
                // Arrange
                jest.spyOn(winston, "createLogger");
                const fakePrintF = jest.fn();
                jest.spyOn(winston.format, "combine", "get").mockReturnValue(
                    jest.fn(),
                );
                jest.spyOn(winston.format, "printf", "get").mockReturnValue(
                    fakePrintF,
                );

                // Act
                createLogger({mode: Runtime.Development, level: "silly"});
                const result = fakePrintF.mock.calls[0][0]({
                    level: "debug",
                    message: "MESSAGE",
                    ...otherMetadata,
                });

                // Assert
                expect(result).toMatchSnapshot();
            },
        );
    });

    describe("during production", () => {
        it("should write to the console", () => {
            // Arrange
            const mockCreateLogger = jest.spyOn(winston, "createLogger");

            // Act
            createLogger({mode: Runtime.Development, level: "silly"});
            const {transports} = mockCreateLogger.mock.calls[0][0];

            // Assert
            expect(transports).toBeInstanceOf(winston.transports.Console);
        });

        it.each([undefined, {other: "metadata"}, {}])(
            "should format log messages to include metadata when present (%s)",
            (otherMetadata) => {
                // Arrange
                jest.spyOn(winston, "createLogger");
                const fakePrintF = jest.fn();
                jest.spyOn(winston.format, "combine", "get").mockReturnValue(
                    jest.fn(),
                );
                jest.spyOn(winston.format, "printf", "get").mockReturnValue(
                    fakePrintF,
                );

                // Act
                createLogger({mode: Runtime.Development, level: "silly"});
                const result = fakePrintF.mock.calls[0][0]({
                    level: "debug",
                    message: "MESSAGE",
                    ...otherMetadata,
                });

                // Assert
                expect(result).toMatchSnapshot();
            },
        );
    });

    it("should log creation parameters to created logger", () => {
        // Arrange
        const debugSpy = jest.fn();
        jest.spyOn(winston, "createLogger").mockReturnValue({
            debug: debugSpy,
        });

        // Act
        createLogger({mode: Runtime.Test, level: "info"});

        // Assert
        expect(debugSpy).toHaveBeenCalledWith(
            "Created logger (Level=info Mode=test)",
        );
    });
});
