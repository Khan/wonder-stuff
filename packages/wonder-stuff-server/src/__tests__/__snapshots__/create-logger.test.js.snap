// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#createLogger during development should format log messages to include metadata when present ({ other: 'metadata' }) 1`] = `
"debug: MESSAGE {
    \\"other\\": \\"metadata\\"
}"
`;

exports[`#createLogger during development should format log messages to include metadata when present ({}) 1`] = `"debug: MESSAGE"`;

exports[`#createLogger during development should format log messages to include metadata when present (undefined) 1`] = `"debug: MESSAGE"`;

exports[`#createLogger during production should format log messages to include metadata when present ({ other: 'metadata' }) 1`] = `
"debug: MESSAGE {
    \\"other\\": \\"metadata\\"
}"
`;

exports[`#createLogger during production should format log messages to include metadata when present ({}) 1`] = `"debug: MESSAGE"`;

exports[`#createLogger during production should format log messages to include metadata when present (undefined) 1`] = `"debug: MESSAGE"`;
